xcdat/temporal_avg.py

```diff
@@ -0,0 +1,844 @@
+        with xr.set_options(keep_attrs=True):
+            departures[da_data_var.name] = data_var_grouped - da_data_var
+
+        return departures
+
+    def normalize_time(self, dataset: xr.Dataset) -> xr.Dataset:
+        """Normalizes time coordinates as the midpoint between time bounds.
+
+        Parameters
+        ----------
+        dataset : xr.Dataset
+            The Dataset with the original time coordinates.
+
+        Returns
+        -------
+        xr.Dataset
+            The Dataset with the normalized time coordinates.
+        """
+        time_key = self.normalized_time.name
+        ds = dataset.assign_coords({time_key: (time_key, self.normalized_time.data)})
+        ds[time_key].attrs = self.normalized_time.attrs
+
+        return ds
+
+    def _set_obj_attrs(
+        self,
+        mode: Union[Mode, Literal["departure"]],
+        freq: Frequency,
+        is_weighted: bool,
+        djf_type: DJFType,
+    ):
+        """Validate user inputs and set them as object attributes.
+
+        Parameters
+        ----------
+        mode : Union[Mode, Literal["departure]]
+            The type of mode for temporal averaging.
+        freq : Frequency
+            The frequency of time to group by.
+        is_weighted : bool
+            Perform weighted averages.
+        djf_type : DJFType
+            Whether the DJF season is "cont" (continuous with previous year
+            December) or "discont" (discontinuous with same year Dec).
+
+        Raises
+        ------
+        KeyError
+            If the Dataset does not have a "time" dimension.
+        ValueError
+            If an incorrect ``freq`` arg was passed.
+        ValueError
+            If an incorrect ``djf_type`` arg was passed.
+        """
+        # "departure" is not included in `Mode` because it is only passed
+        # internally  though `.departure()`
+        if mode not in list(MODES) + ["departure"]:
+            modes = ", ".join(f'"{word}"' for word in MODES)
+            raise ValueError(
+                f"Incorrect `mode` argument. Supported modes include: " f"{modes}."
+            )
+
+        freqs_keys = FREQ_GROUPBY_MAP[mode].keys()
+        if freq not in freqs_keys:
+            freqs = ", ".join(f'"{word}"' for word in FREQUENCIES)
+            raise ValueError(
+                f"Incorrect `freq` argument. Supported frequencies for {mode} "
+                f"include: {freqs}."
+            )
+
+        if djf_type not in DJF_TYPES:
+            djf_types = ", ".join(f'"{word}"' for word in DJF_TYPES)
+            raise ValueError(
+                f"Incorrect `djf_type` argument. Supported DJF types include: {djf_types}"
+            )
+
+        self.mode = mode
+        self.freq = freq
+        self.is_weighted = is_weighted
+        self.djf_type = djf_type
+
+    def _handle_custom_seasons(self, custom_seasons: List[List[str]]) -> List[str]:
+        """Validates the user-specified list of custom seasons.
```

Comment:
Change description to `Validates the list of custom seasons and handles creating them.'